internal/docker/builder.go

```diff
@@ -371,51 +382,77 @@ func getCaVolume(docker *client.Client, ctx context.Context) (map[string]struct{
 			// We did not find a volume. This container might be created without a volume,
 			// or CI=true is passed but we are not running in a container.
 			// todo: log that we do not provide a CA volume mount?
-			return nil, nil, nil
+			return caVolume, caMount, nil
 		} else {
-			caVolume = map[string]struct{}{
-				"/ca": {},
-			}
-			caMount = []mount.Mount{
-				{
-					Type:   mount.TypeVolume,
-					Source: volumeName,
-					Target: "/ca",
-				},
+			caVolume = "/ca"
+			caMount = mount.Mount{
+				Type:   mount.TypeVolume,
+				Source: volumeName,
+				Target: "/ca",
 			}
 		}
 	} else {
 		// When not in CI, our CA cert is placed in the current working dir.
 		// We bind mount this directory to all homeserver containers.
 		cwd, err := os.Getwd()
 		if err != nil {
-			return nil, nil, err
+			return caVolume, caMount, err
 		}
 		caCertificateDirHost := path.Join(cwd, "ca")
 		if _, err := os.Stat(caCertificateDirHost); os.IsNotExist(err) {
 			err = os.Mkdir(caCertificateDirHost, 0770)
 			if err != nil {
-				return nil, nil, err
+				return caVolume, caMount, err
 			}
 		}
-		caMount = []mount.Mount{
-			{
-				Type:   mount.TypeBind,
-				Source: path.Join(cwd, "ca"),
-				Target: "/ca",
-			},
+
+		caMount = mount.Mount{
+			Type:   mount.TypeBind,
+			Source: path.Join(cwd, "ca"),
+			Target: "/ca",
 		}
 	}
 	return caVolume, caMount, nil
 }
 
+// getAppServiceVolume returns the correct mounts and volumes for providing the `/appservice` directory to homeserver containers
+// containing application service registration files to be used by the homeserver
+func getAppServiceVolume(docker *client.Client, ctx context.Context) (string, mount.Mount, error) {
+	asVolume, err := docker.VolumeCreate(context.Background(), volume.VolumesCreateBody{
+		//Driver:     "overlay2",
+		DriverOpts: map[string]string{},
+		Name:       "appservices",
+	})
+
+	asMount := mount.Mount{
+		Type:   mount.TypeVolume,
+		Source: asVolume.Name,
+		Target: "/appservices",
+	}
+
+	return "/appservices", asMount, err
+}
+
+func generateASRegistrationYaml(as b.ApplicationService) string {
+	return fmt.Sprintf("id: %s\n", as.ID) +
+		fmt.Sprintf("hs_token: %s\n", as.HSToken) +
+		fmt.Sprintf("as_token: %s\n", as.ASToken) +
+		fmt.Sprintf("url: '%s'\n", as.URL) +
+		fmt.Sprintf("sender_localpart: %s\n", as.SenderLocalpart) +
+		fmt.Sprintf("rate_limited: %v\n", as.RateLimited) +
+		"namespaces:\n" +
```

Comment:
You can use literal syntax via `

Reply:
String literal docs: https://golang.org/ref/spec#String_literals

---

Initially, I used a multiline backtick raw string but the `fmt.Sprintf` on it was pretty weird since I needed to make sure the order was correct over a list of 6 variables at the end. Seems pretty easy to get it wrong whenever touching it.

Also, since whitespaces matters, it looks out of place to have everything smashed against the beginning of the line and doesn't match the surrounding indentation of the code.

Is there a cleaner way to substitute variables (string interpolation)?

Reply:
Hmm, good points. Nope, I think the combo of needing to inject vars and whitespace are good enough reasons not to use string literals.